File: esj-rules/src/main/java/de/bsnsoft/esj/rules/MessageTemplate.java
package de.bsnsoft.esj.rules;

import de.bsnsoft.esj.SemanticPath;
import de.bsnsoft.esj.SemanticType;
import java.util.ArrayList;
import java.util.List;
import java.util.Map;

/**
 * The message of a rule, with the places the values go.
 *
 * <p>A finding that says only that BR-CO-10 failed makes the reader open the invoice and do
 * the arithmetic. A finding that says which figure was expected, which was there and where
 * both live makes the reader fix the invoice. So a message is a template, and the values it
 * shows are read out of the document when the rule fires.
 *
 * <p>Four placeholders, and each of them is resolved when the pack is compiled, so a message
 * that names a business term nobody has is a defect of the pack rather than a surprise in a
 * report:
 *
 * <ul>
 *   <li>{@code {/BG-22/BT-106}} — the value at that path, relative to the rule's context,
 *       escaped as {@code SPEC.md} section 9.5 requires and cut to the excerpt length of
 *       {@link Texts}, so that a message quotes enough of a value to be recognised and
 *       never the whole of a large one. A value that is not there reads
 *       {@code (absent)}.</li>
 *   <li>{@code {@/BG-22/BT-106}} — the path itself, as the document addresses it, so that a
 *       reader can go to it.</li>
 *   <li>{@code {$name}} — an expression the rule bound under that name, which is how a sum
 *       or a difference gets into a message.</li>
 *   <li>{@code {.}} — the business group instance the rule is looking at, {@code /} for the
 *       document.</li>
 * </ul>
 *
 * <p>A literal brace is written twice.
 */
final class MessageTemplate {

    /** One piece of a message: a constant, or something read when the rule fires. */
    @FunctionalInterface
    private interface Part {
        String render(Evaluation evaluation, SemanticPath base);
    }

    private final List<Part> parts;

    private MessageTemplate(List<Part> parts) {
        this.parts = parts;
    }

    /**
     * Compiles a template.
     *
     * @param template the template as the rule file writes it
     * @param compiler the compiler, for the paths in the template
     * @param scope    the context of the rule
     * @param bindings the expressions the rule bound, already compiled
     * @param where    what is being compiled, for the message of a failure
     * @return the compiled template
     * @throws RulePackException if the template names something that cannot be resolved
     */
    static MessageTemplate compile(String template, Compiler compiler, Compiler.Scope scope,
                                   Map<String, Compiler.Compiled> bindings, String where) {
        List<Part> parts = new ArrayList<>();
        StringBuilder literal = new StringBuilder();
        int i = 0;
        while (i < template.length()) {
            char c = template.charAt(i);
            if (c == '}' && i + 1 < template.length() && template.charAt(i + 1) == '}') {
                literal.append('}');
                i += 2;
                continue;
            }
            if (c != '{') {
                literal.append(c);
                i++;
                continue;
            }
            if (i + 1 < template.length() && template.charAt(i + 1) == '{') {
                literal.append('{');
                i += 2;
                continue;
            }
            int end = template.indexOf('}', i + 1);
            if (end < 0) {
                throw new RulePackException(where + ": the message opens a placeholder at "
                        + i + " and never closes it");
            }
            if (literal.length() > 0) {
                String text = literal.toString();
                parts.add((evaluation, base) -> text);
                literal.setLength(0);
            }
            parts.add(placeholder(template.substring(i + 1, end), compiler, scope, bindings, where));
            i = end + 1;
        }
        if (literal.length() > 0) {
            String text = literal.toString();
            parts.add((evaluation, base) -> text);
        }
        return new MessageTemplate(List.copyOf(parts));
    }

    private static Part placeholder(String name, Compiler compiler, Compiler.Scope scope,
                                    Map<String, Compiler.Compiled> bindings, String where) {
        if (name.equals(".")) {
            return (evaluation, base) -> base.isRoot() ? "/" : base.toString();
        }
        if (name.startsWith("$")) {
            Compiler.Compiled bound = bindings.get(name.substring(1));
            if (bound == null) {
                throw new RulePackException(where + ": the message names " + name
                        + ", which the rule does not bind");
            }
            return (evaluation, base) -> bound.expression().evaluate(evaluation, base).display();
        }
        if (name.startsWith("@")) {
            PathPattern pattern = compiler.singleValuePath(name.substring(1), scope, where);
            return (evaluation, base) -> pattern.absoluteText(base);
        }
        if (name.startsWith("/")) {
            PathPattern pattern = compiler.singleValuePath(name, scope, where);
            SemanticType type = compiler.datatypeOf(pattern, where);
            return (evaluation, base) -> evaluation.read(pattern, base, type).display();
        }
        throw new RulePackException(where + ": {" + name + "} is not a placeholder;"
                + " a placeholder is a path, a path after @, a bound name after $, or a dot");
    }

    /**
     * Renders the message.
     *
     * @param evaluation the run
     * @param base       the business group instance the rule is evaluated at, or the root
     * @return the message
     * @throws Undecided if a value the message reads does not spell what its semantic data
     *                   type requires
     */
    String expand(Evaluation evaluation, SemanticPath base) {
        StringBuilder message = new StringBuilder();
        for (Part part : parts) {
            message.append(part.render(evaluation, base));
        }
        return message.toString();
    }
}

File: bindings/csharp/En16931.SemanticJson/Rules/MessageTemplate.cs
using System;
using System.Collections.Generic;
using System.Globalization;
using System.Text;

namespace En16931.SemanticJson.Rules;

/// <summary>
/// What a finding says, with the four placeholders of <c>rules/README.md</c> resolved when
/// the pack is compiled rather than when a rule fails.
/// </summary>
/// <remarks>
/// A finding that says only that a rule failed makes the reader open the invoice and do the
/// arithmetic. <c>{/BG-22/BT-106}</c> shows the value at that path, <c>{@/BG-22/BT-106}</c>
/// the path itself, <c>{$name}</c> an expression the rule bound under that name and
/// <c>{.}</c> the business group instance the rule is looking at. A literal brace is written
/// twice, and a message that names a term nobody has is a defect of the pack, found once.
/// </remarks>
internal sealed class MessageTemplate
{
    private delegate string Part(Evaluation evaluation, SemanticPath basePath);

    private readonly List<Part> _parts;

    private MessageTemplate(List<Part> parts)
    {
        _parts = parts;
    }

    internal static MessageTemplate Compile(
        string template,
        Compiler compiler,
        Scope scope,
        IReadOnlyDictionary<string, Compiled> bindings,
        string where)
    {
        List<Part> parts = new();
        StringBuilder literal = new();
        int at = 0;
        while (at < template.Length)
        {
            char character = template[at];
            if (character == '}' && at + 1 < template.Length && template[at + 1] == '}')
            {
                literal.Append('}');
                at += 2;
                continue;
            }

            if (character != '{')
            {
                literal.Append(character);
                at++;
                continue;
            }

            if (at + 1 < template.Length && template[at + 1] == '{')
            {
                literal.Append('{');
                at += 2;
                continue;
            }

            int end = template.IndexOf('}', at + 1);
            if (end < 0)
            {
                throw new RulePackException(where + ": the message opens a placeholder at "
                    + at.ToString(CultureInfo.InvariantCulture) + " and never closes it");
            }

            if (literal.Length > 0)
            {
                string text = literal.ToString();
                parts.Add((evaluation, basePath) => text);
                literal.Clear();
            }

            parts.Add(Placeholder(template.Substring(at + 1, end - at - 1), compiler, scope, bindings, where));
            at = end + 1;
        }

        if (literal.Length > 0)
        {
            string text = literal.ToString();
            parts.Add((evaluation, basePath) => text);
        }

        return new MessageTemplate(parts);
    }

    private static Part Placeholder(
        string name,
        Compiler compiler,
        Scope scope,
        IReadOnlyDictionary<string, Compiled> bindings,
        string where)
    {
        if (name == ".")
        {
            return (evaluation, basePath) => basePath.IsRoot ? "/" : basePath.Text;
        }

        if (name.StartsWith("$", StringComparison.Ordinal))
        {
            if (!bindings.TryGetValue(name.Substring(1), out Compiled? bound))
            {
                throw new RulePackException(
                    where + ": the message names " + name + ", which the rule does not bind");
            }

            return (evaluation, basePath) => bound.Expression(evaluation, basePath).Display();
        }

        if (name.StartsWith("@", StringComparison.Ordinal))
        {
            PathPattern pattern = compiler.SingleValuePath(name.Substring(1), scope, where);
            return (evaluation, basePath) => pattern.AbsoluteText(basePath);
        }

        if (name.StartsWith("/", StringComparison.Ordinal))
        {
            PathPattern pattern = compiler.SingleValuePath(name, scope, where);
            SemanticType type = compiler.DatatypeOf(pattern, where);
            return (evaluation, basePath) => evaluation.Read(pattern, basePath, type).Display();
        }

        throw new RulePackException(where + ": {" + name + "} is not a placeholder; a placeholder"
            + " is a path, a path after @, a bound name after $, or a dot");
    }

    /// <summary>Returns the message a rule instance produces.</summary>
    /// <param name="evaluation">the run</param>
    /// <param name="basePath">the instance the rule is evaluated at</param>
    /// <returns>the message</returns>
    internal string Expand(Evaluation evaluation, SemanticPath basePath)
    {
        StringBuilder message = new();
        foreach (Part part in _parts)
        {
            message.Append(part(evaluation, basePath));
        }

        return message.ToString();
    }
}
